src/calendar_duration.rs

```diff
@@ -0,0 +1,294 @@
+///
+/// **Note**: The distinction between *minutes and seconds* components and only a *seconds*
+/// component prepares the `CalendarDate` type for working with leap seconds accurately.
+/// However they currently behave the same because full support for leap seconds is not yet
+/// implemented in chrono.
+///
+/// | component           | limit                    |
+/// |---------------------|--------------------------|
+/// | months              | `u32::MAX`               |
+/// | days                | `u32::MAX`               |
+/// | minutes and seconds | `u64::MAX >> 8` and `60` |
+/// | seconds             | `u64::MAX >> 2`          |
+/// | nanoseconds         | `999_999_999`            |
+///
+/// # Examples
+///
+/// ```
+/// # use chrono::CalendarDuration;
+/// // Encoding 1½ year as a duration of a year and 6 months:
+/// let _duration = CalendarDuration::new().with_years_and_months(1, 6).unwrap();
+///
+/// // Encoding 1½ year as a duration of 12 months and 183 days (366 / 2):
+/// let _duration = CalendarDuration::new().with_months(12).with_days(183);
+///
+/// // Encoding 1½ year as a duration of 548 days (365 + 366 / 2):
+/// let _duration = CalendarDuration::new().with_days(548);
+///
+/// // Encoding 1½ year as a duration in seconds:
+/// let _duration = CalendarDuration::new().with_seconds(548 * 24 * 60 * 60).unwrap();
+/// ```
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct CalendarDuration {
+    // Components with a nominal duration
+    months: u32,
+    days: u32,
+    // Components with an accurate duration
+    mins_and_secs: MinutesAndSeconds,
+    nanos: u32,
+}
+
+impl CalendarDuration {
+    /// Create a new duration initialized to `0`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use chrono::CalendarDuration;
+    /// // Duration in calendar months and days.
+    /// let _ = CalendarDuration::new().with_months(15).with_days(5);
+    /// // Duration in calendar days, and 3 hours.
+    /// let _ = CalendarDuration::new().with_days(5).with_hms(3, 0, 0).unwrap();
+    /// // Duration that will precisely count the number of seconds, including leap seconds.
+    /// let _ = CalendarDuration::new().with_seconds(23_456).unwrap().with_millis(789).unwrap();
+    /// ```
+    #[must_use]
+    pub const fn new() -> Self {
+        Self {
+            months: 0,
+            days: 0,
+            mins_and_secs: expect(MinutesAndSeconds::from_seconds(0), "in range"),
+            nanos: 0,
+        }
+    }
+
+    /// Set the months component of this duration to `months`.
+    #[must_use]
+    pub const fn with_months(mut self, months: u32) -> Self {
+        self.months = months;
+        self
+    }
+
+    /// Set the months component of this duration to `years * 12 + months`.
+    ///
+    ///  # Errors
+    ///
+    /// Returns `None` if the value of the months component would be out of range.
+    pub const fn with_years_and_months(mut self, years: u32, months: u32) -> Option<Self> {
```

Comment:
I'd prefer to skip this one for now, easy enough to implement from the outside.

Reply:
I hope to have these eventually. Not only for convenience, but to match the components ISO 8601 defines.